batch/batch/throttler.py

```diff
@@ -13,20 +14,34 @@ def __init__(self, queue_size, max_pods, parallelism=1):
         self.pending_pods = set()
         self.created_pods = set()
 
-        for _ in range(parallelism):
-            asyncio.ensure_future(self._create_pod())
+        workers = [asyncio.ensure_future(self._create_pod())
+                   for _ in range(parallelism)]
+
+        while True:
+            failed, pending = asyncio.wait(workers, return_when=asyncio.FIRST_COMPLETED)
+            for fut in failed:
+                err = fut.exception()
+                assert err is not None
+                err_msg = '\n'.join(
+                    traceback.format_exception(type(err), err, err.__traceback__))
+                log.error(f'restarting failed worker: {err} {err_msg}')
+                pending.append(asyncio.ensure_future(self._create_pod()))
+            workers = pending
 
     async def _create_pod(self):
         while True:
             await self.semaphore.acquire()
 
-            job = await self.queue.get()
-            pod_name = job._pod_name
+            try:
+                job = await self.queue.get()
+                pod_name = job._pod_name
 
-            if pod_name not in self.pending_pods:
-                log.info(f'pod {pod_name} was deleted before it was created, ignoring')
+                if pod_name not in self.pending_pods:
+                    log.info(f'pod {pod_name} was deleted before it was created, ignoring')
+                    self.semaphore.release()
+                    return
+            except:
                 self.semaphore.release()
```

Comment:
don't you need another return here? Also, the try except should probably be around _create_pod below instead?

Reply:
oops I meant to re-raise, yeah I'll include the create_pod